README.md
# VWS JS library

Node module for web-scraping on torrent video websites

Mainly uses a module called [cheerio](https://github.com/cheeriojs/cheerio) by Matthew Mueller which implements a subset of jQuery specifically designed for server use.

Currently websites scraped list:

- [pctmix.org](https://pctmix.org/)
- [dontorrent.org](https://dontorrent.org)

Implemented node modules for search on internet movie databases as:

- [omdb](http://www.omdbapi.com/)
- [tmdb](https://www.themoviedb.org/)

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes.

### Prerequisites

You'll need [Node.js](https://nodejs.org/es/) installed on your computer in order to build this library

```sh
$git clone https://github.com/rvillamil/vws-js-lib
$cd vws-js-lib
$npm install
```

## Running the tests

 For running the automated tests for this system, with [Mocha](https://mochajs.org) Javascript test framework

```sh
$npm test
```

or indiviual test by description

```sh
$npm test -- --grep "crawler"
```

## Release version

I user Github actions + [semantic relases plugin](https://github.com/semantic-release/semantic-release) on push on master. 

**Warning! Version on package.json is automatic. Don't change**

It Requieres commit message:

| Commit message                                                                                                                                                                                   | Release type               |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | -------------------------- |
| `fix(pencil): stop graphite breaking when too much pressure applied`                                                                                                                             | Patch Release              |
| `feat(pencil): add 'graphiteWidth' option`                                                                                                                                                       | ~~Minor~~ Feature Release  |
| `perf(pencil): remove graphiteWidth option`<br><br>`BREAKING CHANGE: The graphiteWidth option has been removed.`<br>`The default graphite width of 10mm is always used for performance reasons.` | ~~Major~~ Breaking Release |


Example (on master):

```sh
$git ci -a -m "feat(ci): Remove step in CI"
```

..then

```sh
$git push
```

When CI ends, then, there is a release version generated on:

- https://www.npmjs.com/package/@vws-js-lib/crawler
- https://github.com/rvillamil/vws-js-lib/releases


## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details.


CHANGELOG.md
## [4.1.0] - 2021-07-12

- Several Bug fixed on pctmix1 parsing

## [4.0.6] - 2021-05-31

- Bug fixed on pctmix1 parsing

## [4.0.5] - 2021-05-31

- pctmix1 support

## [4.0.4] - 2021-03-02

- Bug fixed parsing year on pctmix
- Other minor bug fixed

## [4.0.3] - 2021-02-25

- New URL to download
- Minor bug fixed
  
## [4.0.2] - 2020-12-02

- pctmix.com site replace descargas2020 site

## [4.0.1] - 2020-08-02

- Bug fixed on films with title between '[]'
- Minor bugs on parsing

## [4.0.0] - 2019-12-13

- New dontorrent.com crawler!
- New dontorrent.com searcher!
- Renaming tool 'dbutils' to 'favorites-tool'

## [3.0.1] - 2019-11-23

### Changed

- Upgrading 3rdParty libraries
- Bug fixed on title parsing
- ESLint minor bugs fixed
- Minor bugs on test

## [3.0.0] - 2019-07-06

### Changed

- Change the portal to parse: Now is 'pctmix.org'
- Refactor code

## [2.1.4] - 2019-02-24

### Changed

- minor bug fixed

## [2.1.3] - 2019-02-24

### Changed

- tumejortorrent site, has been replaced by newpct video web site

## [2.1.2] - 2019-02-15

### Changed

- Fix error. tumejortorrent has changed to 'https' url

## [2.1.1] - 2018-12-16

### Changed

- Fix error on load cover from tumejortorrent
- Fix error on parse tvshow title from tumejortorrent
- Fix test

## [2.1.0] - 2018-10-27

### Added

- Now persist all tvshows information about the showcollection in favorites database
- New favorites tool management console: `utils/dbutils`

## [2.0.2] - 2018-10-07

### Changed

- Fix problem on linux distributions

## [2.0.1] - 2018-10-07

### Added

- New node tool for search in OMDB and TMDB : `utils/xmdb.js`

### Changed

- Minor bug fixed

- OMDB and TMDB search enhacements and results

## [2.0.0] - 2018-04-21

### Added

- API enhacements: new objects. Incompatible changes with previous version
- New LinkChained object. Wrapper for URL Links
- New Showcollection object. e.g: TVShow sessions are show collection
- New Database support: nedb
- New FavoriteRepository object, with 'showCollection' persistence support
- Test enhacements

## [1.0.0] - 2018-03-30

### Added

- New API Facade 'crawler'
- New API for tumejortorrent crawler
- New TMDB support
- New OMDB API Support
- New data: Original title, year, tmdbRating, imdbRating, rottenTomatoes
- Other minor bug fixed

## [0.0.3] - 2018-03-16

### Changed

- Tests fixed

- Minor bugs fixed

## [0.0.2] - 2018-03-10

### Changed

- API enhanced

### Added

- Get all torrent shows on sync mode

## [0.0.1] - 2018-02-20

### Added

- First big commit


lib/tmdb.js
//
// See API: https://www.themoviedb.org/documentation/api
//   - Example: https://api.themoviedb.org/3/search/movie?api_key=f8f6fe20a118a8362fd4e17fa1313703&language=en-US&query=Ghost&page=1&include_adult=false&year=1990
//   - Request Limits: https://developers.themoviedb.org/3/getting-started/request-rate-limiting
//

// NPM modules
const rp = require('request-promise')
const Show = require('./model/show')

// Constants
const TMDB_API_KEY = 'f8f6fe20a118a8362fd4e17fa1313703' // Example: https://api.themoviedb.org/3/movie/550?api_key=f8f6fe20a118a8362fd4e17fa1313703
const URL_BASE = 'http://api.themoviedb.org/3/'

//
// Export my NPM functions
//
exports.searchShow = function (title, year, kind, debug = false) {
    return _doSearch(title, year, kind, debug)
}

// ----------------------------------------------------------------------------
//
// Private functions
//
// @see API: https://developers.themoviedb.org/3/search/search-movies
function _doSearch(title, year, kind = 'movie', debug = false) {
    var requestOtions = {
        uri: `${URL_BASE}search/${kind}`,
        qs: {
            api_key: TMDB_API_KEY,
            query: title,
            year: year,
            include_adult: 'false',
            page: '1',
            language: 'es-ES',
        },
        json: true, // Automatically parses the JSON string in the response
    }

    return rp(requestOtions)
        .then(function (response) {
            if (debug) {
                console.log('Response from TMDB: ' + JSON.stringify(response))
            }
            var show = new Show()
            show.title = title
            if (response.total_results != 0) {
                var first = _getExactTitleResult(show.title, response, kind)
                show.sinopsis = first.overview
                show.originalTitle = first.original_title
                show.releaseDate = first.release_date
                if (show.releaseDate) {
                    show.year = first.release_date.split('-')[0]
                }
                if (first.vote_average > 0) {
                    show.tmdbRating = first.vote_average
                }
                if (first.poster_path) {
                    show.urlwithCover = `http://image.tmdb.org/t/p/w185/${first.poster_path}`
                }
                show.error = 0
            } else {
                show.error = 1
            }

            return show
        })
        .catch((err) => {
            console.error('Response from TMDB: ' + err)
            return err
        })
}

/**
 * Obtiene el primer resultado mas cercano al año en curso. La respuesta viene ordenada por popularidad
 * @param {*} response
 */
/*
function _getNearestDateResult(response, kind) {
    var nearestShowByYear = response.results[0]
    var today = new Date()
    var yyyy = today.getFullYear()

    for (var i = 0; i < response.total_results; ++i) {
        var releaseDateShow = null
        if (kind==='tv') {
            releaseDateShow = new Date(response.results[i].first_air_date)
        }
        if (kind==='movie') {
            releaseDateShow = new Date(response.results[i].release_date)
        }
        var yearShow = releaseDateShow.getFullYear()
        if (yearShow <= yyyy) {
            nearestShowByYear = response.results[i]
            return nearestShowByYear
        }
    }
    return nearestShowByYear
}
*/

/**
 * Obtiene el resultado para el que coincide el titulo exactamente
 * @param {*} response
 */
function _getExactTitleResult(title, response, kind) {
    var nearestShowByYear = response.results[0]
    
    for (var i = 0; i < response.total_results; ++i) {
        //console.log (response.results[i].title)    
        if ( response.results[i].title.toUpperCase() === title.toUpperCase() ){
            nearestShowByYear = response.results[i]
            return nearestShowByYear
        }
    }
    return nearestShowByYear
}

// JSON return from TMDB ...
/*
{
    "page":1,
    "total_results":2,
    "total_pages":1,
    "results":[
       {
          "vote_count":8428,
          "id":11,
          "video":false,
          "vote_average":8.1,
          "title":"La guerra de las galaxias. Episodio IV: Una nueva esperanza",
          "popularity":45.430126,
          "poster_path":"/8ae71OAm6XwnvakAx6rYa1Lo5qD.jpg",
          "original_language":"en",
          "original_title":"Star Wars",
          "genre_ids":[
             12,
             28,
             878
          ],
          "backdrop_path":"/4iJfYYoQzZcONB9hNzg0J0wWyPH.jpg",
          "adult":false,
          "overview":"La princesa Leia, líder del movimiento rebelde que desea reinstaurar la República en la galaxia en los tiempos ominosos del Imperio, es capturada por las malévolas Fuerzas Imperiales, capitaneadas por el implacable Darth Vader, el sirviente más fiel del emperador. El intrépido Luke Skywalker, ayudado por Han Solo, capitán de la nave espacial \"El Halcón Milenario\", y los androides, R2D2 y C3PO, serán los encargados de luchar contra el enemigo y rescatar a la princesa para volver a instaurar la justicia en el seno de la Galaxia.",
          "release_date":"1977-05-25"
       },
       {
          "vote_count":15,
          "id":72694,
          "video":false,
          "vote_average":6.4,
          "title":"The Making of 'Star Wars'",
          "popularity":2.124669,
          "poster_path":"/WgsInarGqxj5a33H1oO1IJACDU.jpg",
          "original_language":"en",
          "original_title":"The Making of 'Star Wars'",
          "genre_ids":[
             99
          ],
          "backdrop_path":"/eT9liR62vzdqrktB7wxiYyqeDHA.jpg",
          "adult":false,
          "overview":"",
          "release_date":"1977-09-16"
       }
    ]
 }
 */
